### PR TITLE
[Issue 2461][pulsar-client-cpp] Modified CMake files and source to enable compilation on Windows

### LINKED ISSUE #2461: windows server support needed!
for some reason,we must stay windows platform.
 compiling  pulsar-cpp always give errors.
any plan offering support for windows? thanks!

### PR DESCRIPTION
**Pull request by heronr**
Fixes #2461

### Motivation

My motivation was to be able to use the pulsar cpp client on Windows systems. 

### Modifications

There are a number of modifications I needed to make to enable Windows compilation
* Extensive rework of multiple CMakeLists.txt files
* The creation of a PULSAR_PUBLIC define to define symbol visibility in lieue of #pragma GCC visibility push(default). This is because Windows requires specifying __declspec(dllexport) and __declspec(import) explicitly and does not have a #pragma GCC visibility push(default) analogue.
* all calls to usleep or sleep have been replaced with calls to boost::this_thread::sleep()

### Verifying this change

This change is a trivial rework / code cleanup without any test coverage. However, it does introduce a new platform to test on. It is likely that CI checks will have to be created. 

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
	- It adds a dependency to the [dlfcn-win32](https://github.com/dlfcn-win32/dlfcn-win32) library for Windows builds. 
  - The public API: no
  - The schema: no
  - The default values of configurations: no
  - The wire protocol: no
  - The rest endpoints: no
  - The admin cli options: no
  - Anything that affects deployment: no

### Documentation

  - Does this pull request introduce a new feature?: No
  - The README.md file in pulsar-client-cpp has been updated with instructions for Windows compilation

### CODE REVIEW

File: pulsar-client-cpp/tests/BackoffTest.cc [hunk 1]
```diff
@@ -51,7 +53,7 @@ TEST(BackoffTest, firstBackoffTimerTest) {
     Backoff backoff(milliseconds(100), seconds(60), milliseconds(1900));
     ASSERT_EQ(backoff.next().total_milliseconds(), 100);
     boost::posix_time::ptime firstBackOffTime = PulsarFriend::getFirstBackoffTime(backoff);
-    usleep(300 * 1000);
+    boost::this_thread::sleep(boost::posix_time::milliseconds(300));
```
Comment (merlimat) [inline pulsar-client-cpp/tests/BackoffTest.cc]: We could also `std::this_thread::sleep_for()` instead of boost

File: pulsar-client-cpp/lib/ExecutorService.cc [hunk 2]
```diff
@@ -55,9 +55,14 @@ DeadlineTimerPtr ExecutorService::createDeadlineTimer() {
 }
 
 void ExecutorService::close() {
-    io_service_.stop();
-    work_.reset();
-    worker_.join();
+    // Ensure this service has not already been closed. This is
+    // because worker_.join() is not re-entrant on Windows
+    if (work_)
+    {
```
Comment (merlimat) [inline pulsar-client-cpp/lib/ExecutorService.cc]: Even if Windows related, can you submit this fix as a separate PR so that it gets more visible? 

Also, the `{` on new line will fail on the `make check-format`. You should be able to do `make format` and automatically fix all the formatting (that uses `clang-format` underneath)

**Comment by merlimat** [comment]
Once this is merged, we can start looking in how to enable C++ build validation for Windows. Apache's Jenkins has several Windows worker nodes.

File: pulsar-client-cpp/tests/BackoffTest.cc (hunk 1 above)
Comment (heronr) [reply pulsar-client-cpp/tests/BackoffTest.cc]: That's true and it does mean one less boost lib to link. If that is preferable I can go ahead and make the change.

File: pulsar-client-cpp/lib/Backoff.cc [hunk 3]
```diff
@@ -25,8 +25,11 @@ Backoff::Backoff(const TimeDuration& initial, const TimeDuration& max, const Tim
       max_(max),
       next_(initial),
       mandatoryStopMade_(false),
-      mandatoryStop_(mandatoryStop),
-      randomSeed_(time(NULL)) {}
+      mandatoryStop_(mandatoryStop)
+#ifndef _MSC_VER
```
[the remark this answers is not in the document]
Comment (heronr) [reply pulsar-client-cpp/lib/Backoff.cc]: That's a great idea. I went ahead modified Backoff to use a boost::random::mt19937 object for random number generation.

File: pulsar-client-cpp/lib/ExecutorService.cc (hunk 2 above)
Comment (heronr) [reply pulsar-client-cpp/lib/ExecutorService.cc]: Ok, I took this change out of this PR and into a new one, #4084 

In addition, I realized that I didn't notice the check-format failure because the Windows version of the CMake files were omitting those targets. I re-added them and was able to correct a host of formatting mistakes.

File: pulsar-client-cpp/tests/BackoffTest.cc (hunk 1 above)
Comment (heronr) [reply pulsar-client-cpp/tests/BackoffTest.cc]: I did go ahead and switch everything to std::this_thread::sleep_for as you suggested

Comment (merlimat) [reply pulsar-client-cpp/tests/BackoffTest.cc]: Nice. Thanks!

**Comment by merlimat** [comment]
@heronr There seems to be a linking issue on the test tools (protobuf missing) : https://builds.apache.org/job/pulsar_precommit_cpp/7133/console

**Comment by heronr** [comment]
@merlimat Finally got all the tests passing. Should be good to go.
